src/main/java/it/gov/pagopa/pu/debtpositions/service/create/receipt/InstallmentUpdateServiceImpl.java

```diff
@@ -0,0 +1,87 @@
+package it.gov.pagopa.pu.debtpositions.service.create.receipt;
+
+import it.gov.pagopa.pu.debtpositions.dto.generated.InstallmentStatus;
+import it.gov.pagopa.pu.debtpositions.dto.generated.ReceiptDTO;
+import it.gov.pagopa.pu.debtpositions.exception.custom.NotFoundException;
+import it.gov.pagopa.pu.debtpositions.model.DebtPosition;
+import it.gov.pagopa.pu.debtpositions.model.InstallmentNoPII;
+import it.gov.pagopa.pu.debtpositions.model.InstallmentSyncStatus;
+import it.gov.pagopa.pu.debtpositions.repository.DebtPositionRepository;
+import it.gov.pagopa.pu.organization.dto.generated.Broker;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.stereotype.Service;
+
+import java.util.Set;
+
+@Service
+@Slf4j
+public class InstallmentUpdateServiceImpl implements InstallmentUpdateService{
+
+  public static final Set<InstallmentStatus> NOT_PAID = Set.of(InstallmentStatus.DRAFT, InstallmentStatus.TO_SYNC, InstallmentStatus.UNPAID, InstallmentStatus.EXPIRED);
+
+
+  private final DebtPositionRepository debtPositionRepository;
+
+  InstallmentUpdateServiceImpl(DebtPositionRepository debtPositionRepository) {
+    this.debtPositionRepository = debtPositionRepository;
+  }
+
+  @Override
+  public DebtPosition updateInstallmentStatusOfDebtPosition(InstallmentNoPII installment, Broker broker, ReceiptDTO receiptDTO) {
+    //retrieve debt position
+    DebtPosition debtPosition = debtPositionRepository.findByInstallmentId(installment.getInstallmentId());
+    if(debtPosition == null) {
+      throw new NotFoundException("debt position not found for installment " + installment.getInstallmentId());
+    }
+
+    //update installment status
+    boolean[] primaryInstallmentFound = {false};
+    debtPosition.getPaymentOptions().forEach(paymentOption -> {
+      if (paymentOption.getPaymentOptionId().equals(installment.getPaymentOptionId())) {
+        // current payment option: find this installment
+        paymentOption.getInstallments().stream()
+          .filter(anInstallment -> anInstallment.getInstallmentId().equals(installment.getInstallmentId()))
+          .findAny().ifPresent(primaryInstallment -> {
+            // set status of the found installment to PAID and link it to the receipt
+            log.info("Installment [{}] found for receipt [{}]", primaryInstallment.getInstallmentId(), receiptDTO.getReceiptId());
+            primaryInstallmentFound[0] = true;
+            updateInstallmentFields(primaryInstallment, InstallmentStatus.PAID, broker, receiptDTO.getReceiptId());
+          });
+      } else {
+        // change status of every other installment of different payment options of this debt position to INVALID
+        paymentOption.getInstallments().forEach(anInstallment -> {
+          if (NOT_PAID.contains(anInstallment.getStatus())) {
```

Comment:
why not
```suggestion
          if (InstallmentStatus.UNPAID.equals(anInstallment.getStatus())) {
```

DRAFT and TO_SYNC are not possible
EXPIRED and CANCELED should not be changed

if TO_SYNC and the updateStatusTo is UNPAID, we could think to recoved the TO_SYNC status, changing into INVALID

Reply:
CANCELED is not included in set NOT_PAID.
But in my opinion, EXPIRED should be changed to INVALID, because EXPIRED is not a "final" status, and we want all installment of other payment options to be in a final status (so that there is no way they could be modified anymore).
The same applies to DRAFT (even though I agree with you that if there are installment in DRAFT there is a data coherency issue; but keeping in the set would be on the "safe side").
Regarding TO_SYNC: whatever is the syncStatusTo of the current TO_SYNC I can't see a reason for not replacing the syncStatusTo to INVALID.

Reply:
@robertopasetti what do you think about this?

Reply:
I've tried to summarize the use cases and to identify how to manage the installment status. Here my idea:[
](https://pagopa.atlassian.net/wiki/spaces/SPAC/pages/1420755133/Macchina+Stati+Posizione+Debitoria#Gestione-EXPIRED)

Reply:
ok, so an EXPIRED PO will became INVALID if another PO is PAID or PARTIALLY_PAID